src/rgw/rgw_base_pubsub_rest.cc

```diff
@@ -0,0 +1,205 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "rgw_base_pubsub_rest.h"
+#include "common/dout.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rgw
+
+void RGWPSCreateTopicOp::execute() {
+  op_ret = get_params();
+  if (op_ret < 0) {
+    return;
+  }
+
+  ups = std::make_unique<RGWUserPubSub>(store, s->owner.get_id());
+  op_ret = ups->create_topic(topic_name, dest, topic_arn);
+  if (op_ret < 0) {
+    ldout(s->cct, 1) << "failed to create topic '" << topic_name << "', ret=" << op_ret << dendl;
+    return;
+  }
+  ldout(s->cct, 20) << "successfully created topic '" << topic_name << "'" << dendl;
+}
+
+void RGWPSListTopicsOp::execute() {
+  ups = std::make_unique<RGWUserPubSub>(store, s->owner.get_id());
+  op_ret = ups->get_user_topics(&result);
+  if (op_ret < 0) {
+    ldout(s->cct, 1) << "failed to get topics, ret=" << op_ret << dendl;
+    return;
+  }
+  ldout(s->cct, 20) << "successfully got topics" << dendl;
+}
+
+void RGWPSGetTopicOp::execute() {
+  op_ret = get_params();
+  if (op_ret < 0) {
+    return;
+  }
+  ups = std::make_unique<RGWUserPubSub>(store, s->owner.get_id());
```

Comment:
I'm sure this isn't new, but why is it necessary to use the free store to construct an object with block scope?

Reply:
i think this is mainly because ops don't get a RGWRados* or req_state* until `RGWOp::init(RGWRados*, req_state*, RGWHandler*)` is called, so the RGWPSGetTopicOp constructor can't construct RGWUserPubSub directly. `std::optional` is a good alternative to `std::unique_ptr` for cases like this where you just want to defer construction

Reply:
@cbodley thanks for the suggestion. changing from std::unique_ptr to std::optional
@mattbenjamin can you please clarify your comment? the "topics", which are not tied to a specific bucket, are written using:  ```rgw_put_system_obj(store,...)``` - this is where the "store" object is used